File: scripts/exec_model/forwarder.py
"""`BatchForwarder` — the routing-only category, and its three standard mappings.

`GpuMergePartitions`, `GpuUnion` and `GpuInterleave` never touch rows and never buffer:
they relabel lanes. One driver arm serves all three by cycling `sources_of(p)`, so the
merge's round-robin and the interleave's per-lane child rotation are the same rule
applied to different mappings.
"""

from __future__ import annotations

from abc import ABC, abstractmethod


class BatchForwarder(ABC):
    @abstractmethod
    def out_lanes(self) -> int: ...

    @abstractmethod
    def sources_of(self, out_lane: int) -> list[tuple[int, int]]:
        """The (child index, child lane) pairs feeding `out_lane`, in service order."""


class MergePartitionsForwarder(BatchForwarder):
    """N → 1: lane 0 ← [(0,0), (0,1), …, (0,N-1)], cycled round-robin."""

    def __init__(self, n_in: int):
        self.n_in = n_in

    def out_lanes(self) -> int:
        return 1

    def sources_of(self, out_lane: int) -> list[tuple[int, int]]:
        assert out_lane == 0
        return [(0, lane) for lane in range(self.n_in)]


class UnionForwarder(BatchForwarder):
    """Lane counts sum: output lane k is served by exactly one input lane."""

    def __init__(self, child_lane_counts):
        self.child_lane_counts = list(child_lane_counts)
        self._map: list[tuple[int, int]] = [
            (child, lane)
            for child, count in enumerate(self.child_lane_counts)
            for lane in range(count)
        ]

    def out_lanes(self) -> int:
        return len(self._map)

    def sources_of(self, out_lane: int) -> list[tuple[int, int]]:
        return [self._map[out_lane]]


class InterleaveForwarder(BatchForwarder):
    """Lane p ← [(0,p), (1,p), …, (k-1,p)] — child-major, lane count preserved."""

    def __init__(self, n_children: int, n_lanes: int):
        self.n_children = n_children
        self.n_lanes = n_lanes

    def out_lanes(self) -> int:
        return self.n_lanes

    def sources_of(self, out_lane: int) -> list[tuple[int, int]]:
        return [(child, out_lane) for child in range(self.n_children)]

File: peacockdb-core/src/batch_partitioned/forwarder.rs
//! Batch forwarding: the nodes that renumber lanes without touching rows.
//!
//! `GpuMergePartitions`, `GpuUnion` and `GpuInterleave` are one driver arm over three
//! mappings. A visit to output lane p cycles `sources_of(p)` in listed order, forwarding
//! one batch per visit, skipping sources with nothing queued and retiring those whose
//! producer has finished — the merge's round-robin and the interleave's per-lane child
//! rotation are that same rule.

/// Routes whole batches into a new lane numbering; never touches rows, never buffers.
/// No backends and no `CallStats` — routing is driver work, and a batch's bytes are
/// already accounted as driver-held in flight.
pub trait BatchForwarder {
    /// The (child index, child lane) pairs feeding output lane p, in service order.
    fn sources_of(&self, out_lane: usize) -> Vec<(usize, usize)>;
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Forwarder {
    /// N -> 1: lane 0 takes every lane of the one child, round-robin.
    MergePartitions { n: usize },
    /// Lane counts sum: output lane k is served by exactly one (child, lane).
    Union { lanes: Vec<(usize, usize)> },
    /// Child-major: output lane p is lane p of each child, which is why the inputs must
    /// share a hash distribution.
    Interleave { children: usize, n: usize },
}

impl BatchForwarder for Forwarder {
    fn sources_of(&self, out_lane: usize) -> Vec<(usize, usize)> {
        match self {
            Self::MergePartitions { n } => {
                assert_eq!(out_lane, 0, "GpuMergePartitions has one output lane");
                (0..*n).map(|lane| (0, lane)).collect()
            }
            Self::Union { lanes } => vec![lanes[out_lane]],
            Self::Interleave { children, n } => {
                assert!(
                    out_lane < *n,
                    "output lane {out_lane} is outside the interleave"
                );
                (0..*children).map(|child| (child, out_lane)).collect()
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn merge_takes_every_lane_of_its_one_child() {
        assert_eq!(
            Forwarder::MergePartitions { n: 3 }.sources_of(0),
            vec![(0, 0), (0, 1), (0, 2)]
        );
    }

    #[test]
    fn union_lanes_have_exactly_one_source_each() {
        let union = Forwarder::Union {
            lanes: vec![(0, 0), (0, 1), (1, 0)],
        };
        assert_eq!(union.sources_of(0), vec![(0, 0)]);
        assert_eq!(union.sources_of(2), vec![(1, 0)]);
    }

    #[test]
    fn interleave_serves_lane_p_from_lane_p_of_every_child() {
        let interleave = Forwarder::Interleave { children: 3, n: 2 };
        assert_eq!(interleave.sources_of(1), vec![(0, 1), (1, 1), (2, 1)]);
    }
}
